[WebGL] Calculate correct image transfer size from ArrayBufferView

djg opened the pull request:
#### 592cd219941fef52e1429c31038b2d0c920af544
<pre>
[WebGL] Calculate correct image transfer size from ArrayBufferView
<a href="https://bugs.webkit.org/show_bug.cgi?id=243819">https://bugs.webkit.org/show_bug.cgi?id=243819</a>
rdar://98347913

Reviewed by Kimmo Kinnunen.

Validation was being skipped when pixels was null, which is an optimization
instead of passing a sufficiently large array of all zeroes. Remove the two step
approach to calculating the required bytes and instead return the value from
`validateTexFuncData()` and pass this size to base helper functions.

* LayoutTests/TestExpectations: Enable failing test.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
Unify passing image data span as GCGLSpan&lt;&gt; instead of as separate base ptr +
length.
(WebCore::WebGLRenderingContextBase::computeImageSizeInBytes): Deleted.
(WebCore::WebGLRenderingContextBase::texImageArrayBufferViewHelper):
Unify passing validated image data span as GCGLSpan&lt;&gt; instead of as separate
base ptr + length. Adjustment due to srcOffset moved to validateTexFuncData.
(WebCore::WebGLRenderingContextBase::texImageImpl):
Unify passing image data span as GCGLSpan&lt;&gt; instead of as separate base ptr +
length.
(WebCore::WebGLRenderingContextBase::texImage2DBase):
When using ANGLE, remove extra image size checks as this is handled in
validateTexFuncData.
(WebCore::WebGLRenderingContextBase::texImage3DBase): Deleted
(WebCore::WebGLRenderingContextBase::texSubImage2DBase):
When using ANGLE, remove extra image size checks as this is handled in
validateTexFuncData.
(WebCore::WebGLRenderingContextBase::texSubImage3DBase): Deleted
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
Consolidate calculation of image byte size and don&apos;t skip required checks if
pixels is null. Returns validated span of the data that will be read by texture
specification functions.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
- Remove unneeded texImage helper functions
- Change validateTexFuncData to return optional validated span of image bytes
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
* Source/WebCore/platform/graphics/GrpahicsContextGL.h:
(WebCore::GraphicsContextGL::extractTextureData):
Change pixels to span of pixels instead of ptr.

Canonical link: <a href="https://commits.webkit.org/253543@main">https://commits.webkit.org/253543@main</a>
</pre>

kenrussell reviewed (commented):
Drive-by comment: it'd be ideal if the new test were upstreamed to https://github.com/KhronosGroup/WebGL/tree/main/sdk/tests/conformance2/misc as `invalid-passed-params.html`. Thanks in advance.

(I defer review to @kkinnunen-apple .)

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5675,21 +5661,20 @@ bool WebGLRenderingContextBase::validateArrayBufferType(const char* functionName
     }
 
-    if (!validateSettableTexInternalFormat(functionName, format))
+    if (pixels && !validateSettableTexInternalFormat(functionName, format))
```
The settable tex internal format validation must happen regardless of whether `pixels` is passed or not. This is part of the bug about historical early return that you are fixing.

djg replied:
After scrutinising the use of `validateSettableTexInernalFormat`, I've concluded that the intention is only to call this function when the API is passed image data. This is to reject `gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_SHORT, new Uint16Array(1))` but accept `gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT, 1, 1, 0, gl.DEPTH_COMPONENT, gl.UNSIGNED_SHORT, null)` as specified by `WEBGL_depth_texture` extension.

`validateTexFuncFormatAndType` validates the combination of `internalformat`, `format`, and `type` before reaching this point.

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5675,21 +5661,20 @@ bool WebGLRenderingContextBase::validateArrayBufferType(const char* functionName
         return false;
-    if (!validateArrayBufferType(functionName, type, pixels ? std::optional<JSC::TypedArrayType>(pixels->getType()) : std::nullopt))
+
+    if (pixels && !validateArrayBufferType(functionName, type, pixels ? std::optional<JSC::TypedArrayType>(pixels->getType()) : std::nullopt))
```
`if (pixels && validate(..., pixels ? a: b))` does not make sense on the account that the subexpression is redundant.

The buffer type validation must happen regardless of whether `pixels` is passed or not. This is part of the bug about historical early return that you are fixing. In other words, `validateArrayBufferType` validates both `type` and the array buffer type.

kkinnunen-apple replied:
Currently we have sequence of
- validate parameters and data ptr, adjust them, come up with a new parameter
- adjust parameters even further
- clamp the parameters with the new parameter

you could prototype if we could change that to:
- validate parameters and return the correct parameters

so this would look something like:
`std::optional<GCGLSpan<void> WebGLRenderingContextBase::validateTexFuncData<typename SizeType>(const char* functionName, TexImageDimension texDimension, GCGLsizei width, GCGLsizei height, GCGLsizei depth, GCGLenum format, GCGLenum type, ArrayBufferView* pixels, NullDisposition disposition, GCGLuint srcOffset)`

We would then remove the template parameter SizeType when all ANGLE texture functions would accept a consistent data size parameter.

djg replied:
I've implemented this suggestion and was able to unify and delete some code. 👍

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5698,14 +5683,21 @@ bool WebGLRenderingContextBase::validateTexFuncData(const char* functionName, Te
-    total += skipBytes;
-    if (total.hasOverflowed() || pixels->byteLength() < total) {
+
+    CheckedSize totalImageBytes = totalBytesRequired;
```
This is part of the rabbit hole mentioned elsewhere.
The validate function is used in cases where the size is `GLuint` and in cases size is `GLisize`. You now regress the part which uses `GLuint`

djg replied:
I was trying to standardise on doing the arithmetic with `size_t` and then checking that the result fits into the range of `GLuint` or `GLsizei` at the point we go to one of those APIs. Would you prefer to just always use `GLsizei`?

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5283,20 +5269,20 @@ void WebGLRenderingContextBase::texImageArrayBufferViewHelper(TexImageFunctionID
 #if USE(ANGLE)
     if (functionID == TexImageFunctionID::TexImage3D) {
-        texImage3DBase(target, level, internalformat, width, height, depth, border, format, type, byteLength, data);
+        texImage3DBase(target, level, internalformat, width, height, depth, border, format, type, byteLength, data, &bytesRequired);
```
the bytesRequired is not a out parameter, so there shouldn't be a ptr passed here.

djg replied:
I was using pointer as "optional" parameter.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
```diff
@@ -591,7 +591,7 @@ bool GraphicsContextGL::extractTextureData(unsigned width, unsigned height, GCGL
     unsigned imageSizeInBytes, skipSizeInBytes;
     computeImageSizeInBytes(format, type, width, height, 1, unpackParams, &imageSizeInBytes, nullptr, &skipSizeInBytes);
-    const uint8_t* srcData = static_cast<const uint8_t*>(pixels);
+    const uint8_t* srcData = static_cast<const uint8_t*>(pixels.data);
```
later on, it could make sense to propagate the span downwards to ensure in boundedness.

djg replied:
I agree. I'll do this in a follow-up patch.

djg committed:
[WebGL] Calculate correct image transfer size from ArrayBufferView
https://bugs.webkit.org/show_bug.cgi?id=243819
rdar://98347913

Reviewed by Kimmo Kinnunen.

Validation was being skipped when pixels was null, which is an optimization
instead of passing a sufficiently large array of all zeroes. Remove the two step
approach to calculating the required bytes and instead return the value from
`validateTexFuncData()` and pass this size to base helper functions.

* LayoutTests/TestExpectations: Enable failing test.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::texImageSourceHelper):
Unify passing image data span as GCGLSpan<> instead of as separate base ptr +
length.
(WebCore::WebGLRenderingContextBase::computeImageSizeInBytes): Deleted.
(WebCore::WebGLRenderingContextBase::texImageArrayBufferViewHelper):
Unify passing validated image data span as GCGLSpan<> instead of as separate
base ptr + length. Adjustment due to srcOffset moved to validateTexFuncData.
(WebCore::WebGLRenderingContextBase::texImageImpl):
Unify passing image data span as GCGLSpan<> instead of as separate base ptr +
length.
(WebCore::WebGLRenderingContextBase::texImage2DBase):
When using ANGLE, remove extra image size checks as this is handled in
validateTexFuncData.
(WebCore::WebGLRenderingContextBase::texImage3DBase): Deleted
(WebCore::WebGLRenderingContextBase::texSubImage2DBase):
When using ANGLE, remove extra image size checks as this is handled in
validateTexFuncData.
(WebCore::WebGLRenderingContextBase::texSubImage3DBase): Deleted
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
Consolidate calculation of image byte size and don't skip required checks if
pixels is null. Returns validated span of the data that will be read by texture
specification functions.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
- Remove unneeded texImage helper functions
- Change validateTexFuncData to return optional validated span of image bytes
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
* Source/WebCore/platform/graphics/GrpahicsContextGL.h:
(WebCore::GraphicsContextGL::extractTextureData):
Change pixels to span of pixels instead of ptr.

Canonical link: https://commits.webkit.org/253543@main
